### PR TITLE
launch_ros action to set parameters for all nodes in scope using a global parameter yaml file

### PR DESCRIPTION
**Pull request by adityapande-1995**
This PR aims to add a new ```launch_ros``` action, ```SetParametersFromFile``` which can accept a yaml file with ROS parameters which can be passed to all nodes in scope. 

### The parameter file:
The parameter yaml file must be written in the following format:
```
demo_node_1:
  ros__parameters:
    param_1: 1
    param_2: 5

demo_node_2:
  ros__parameters:
    param_2: 10
    param_4: hello_world
```

### Launch file in xml:
```demo_node_3``` is not in the same scope, so the ```SetParametesFromFile``` action won't affect it.
```
<launch>
  <group>
    <set_parameters_from_file filename='/path/to/param/file.yaml'>
    <node pkg="demo_nodes_cpp" exec="parameter_blackboard" name="demo_node_1">
    <node pkg="demo_nodes_cpp" exec="parameter_blackboard" name="demo_node_2">
  </group>
  <node pkg="demo_nodes_cpp" exec="parameter_blackboard" name="demo_node_3">
</launch>
```

### Equivalent usage in python:
```
from launch import LaunchDescription
from launch_ros.actions import Node, SetParametersFromFile
from launch.actions import GroupAction

def generate_launch_description():
  return LaunchDescription([
     GroupAction(
       actions = [
          SetParametersFromFile('path/to/yaml/file.yaml'),

          Node(
            package = 'demo_nodes_cpp',
            executable = 'parameter_blackboard',
            name = 'demo_node_1'
          ),

          Node(
            package = 'demo_nodes_cpp',
            executable = 'parameter_blackboard',
            name = 'demo_node_2'
          )
        ]
      ),

       Node(
          package = 'demo_nodes_cpp',
          executable = 'parameter_blackboard',
          name = 'demo_node_3'
        )
  ])

```

# TODO 
* Add test cases
* Are ```@property``` methods required for ```SetParametersFromFile``` class ? For e.g. https://github.com/ros2/launch_ros/blob/ef1502ea57f16de388635a86861cb5e84e34c919/launch_ros/launch_ros/actions/set_parameter.py#L76
* Cleaner / better way to extract raw text from ```SomeSubstitutionType``` objects when parsing xml launch files

### CODE REVIEW

**Comment by ivanpauno** [comment]
This is only adding the parameterers to Node actions but not to loaded components, which might be a bit unexpected by users.
Maybe we can fix that after https://github.com/ros2/launch_ros/pull/259 is merged.

File: launch_ros/launch_ros/actions/node.py
Comment (ivanpauno) [inline launch_ros/launch_ros/actions/node.py:428]: this is not the correct way of using substitutions, I think `self.__final_node_name` in L419 is already what you want

File: launch_ros/launch_ros/actions/node.py
Comment (ivanpauno) [inline launch_ros/launch_ros/actions/node.py:436]: you can already use wildcards in node names, e.g. /**.
Maybe it's easier to pass the parameter file to all nodes, `rcl` will already only read the parameters that actually applied to the node

File: launch_ros/launch_ros/actions/set_parameters_from_file.py
Comment (ivanpauno) [inline launch_ros/launch_ros/actions/set_parameters_from_file.py:70]: please check how [normalize_to_list_of_substitutions()](https://github.com/ros2/launch/blob/b59bb6b88b40eaa82870542d93bdfb1a033d05c3/launch/launch/utilities/normalize_to_list_of_substitutions_impl.py#L26) and [perform_substitutions](https://github.com/ros2/launch/blob/b59bb6b88b40eaa82870542d93bdfb1a033d05c3/launch/launch/utilities/perform_substitutions_impl.py#L24) are used in other places.

You need to use the first one in the constructor and the second one in the execute() method

File: launch_ros/launch_ros/actions/node.py
Comment (adityapande-1995) [reply launch_ros/launch_ros/actions/node.py:436]: I added this in case we wanted to pass the same parameter with different values for different nodes. For e.g., ```demo_node_1``` has ```param_2: 5``` and ```demo_node_2``` has ```param_2: 10``` in the example above.

File: launch_ros/launch_ros/actions/node.py
Comment (adityapande-1995) [reply launch_ros/launch_ros/actions/node.py:428]: Also, support for namespaces to be added

File: launch_ros/launch_ros/actions/set_parameters_from_file.py
Comment (adityapande-1995) [reply launch_ros/launch_ros/actions/set_parameters_from_file.py]: fixed

File: launch_ros/launch_ros/actions/__init__.py
Comment (rebecca-butler) [inline launch_ros/launch_ros/actions/__init__.py:39]: nit: prefer alphabetical order (this should be above 'SetRemap')

File: launch_ros/launch_ros/actions/set_parameters_from_file.py
Comment (rebecca-butler) [inline launch_ros/launch_ros/actions/set_parameters_from_file.py:1]: nit: 2021

File: test_launch_ros/test/test_launch_ros/actions/example_parameters_1.yaml
Comment (ivanpauno) [inline test_launch_ros/test/test_launch_ros/actions/example_parameters_1.yaml:1]: nit^N: Now we have a file named, `example_parameters.yaml` and another one named `example_parameters_1.yaml` in the same folder.

I would rather use the same style in both, i.e. `example_parameters_0.yaml` and `example_parameters_1.yaml` (_1 and _2 is also fine).

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameters_from_file.py
Comment (ivanpauno) [inline test_launch_ros/test/test_launch_ros/actions/test_set_parameters_from_file.py:136]: it would be great to add that before merging

**Review by ivanpauno — APPROVED** [review]
LGTM!

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameters_from_file.py
Comment (adityapande-1995) [reply test_launch_ros/test/test_launch_ros/actions/test_set_parameters_from_file.py:136]: ```SetParametersFromFile``` currently only works for nodes and not composable nodes.
As composable nodes and nodes currently use different styles for accepting parameters from yaml files, https://github.com/ros2/launch_ros/pull/259 would need to be closed before I can add that functionality to composable nodes. We could:

1. Add a note to ```SetParametersFromFile``` saying that this won't affect composable nodes, and close this PR. Implement it in a new PR once #259 is merged.
2. Wait for #259 to be merged.

Personally, I think we can proceed with option 1. How should we proceed ? @ivanpauno

Comment (ivanpauno) [reply test_launch_ros/test/test_launch_ros/actions/test_set_parameters_from_file.py:136]: Yes, merging this and iterating after https://github.com/ros2/launch_ros/pull/259 is finished sounds good to me.

**Review by ivanpauno — APPROVED** [review]
LGTM with green CI

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py
Comment (jacobperron) [inline test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py:52]: Why remove this test case?

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameters_from_file.py
Comment (jacobperron) [inline test_launch_ros/test/test_launch_ros/actions/test_set_parameters_from_file.py:1]:
```suggestion
# Copyright 2021 Open Source Robotics Foundation, Inc.
```

**Review by jacobperron — APPROVED** [review]
LGTM! Just a couple minor comments.

File: test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py
Comment (adityapande-1995) [reply test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py:52]: when ``my_param`` is passed again with a different value, it can be reassigned due to ``ros__params`` being a dictionary, with ``my_param`` as a key, and ``3`` as its value. Since we shifted to a list of tuples to preserve order, this would be passed as ``-p my_param:=2 -p another_param:=2 -p my_param:=3`` which would override ``my_param`` anyway.  Thus this test seemed redundant, and won't work as it was originally intended.

Comment (adityapande-1995) [reply test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py:52]: I did try adding a logic to replace the original declaration when it is overridden, but it seemed redundant as ``rcl`` itself is doing it anyway down the line

Comment (jacobperron) [reply test_launch_ros/test/test_launch_ros/actions/test_set_parameter.py:52]: I understand, thanks for explaining; makes sense to me, we can leave it as-is.

**Comment by adityapande-1995** [comment]
CI : 

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15148/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9878/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12837/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15328/)

**Pull request merged**
